Selector by region

ats2008 opened the pull request:
#### PR description:

Module that can take in a track collection and a tracking region , then produce the subset track collection with tracks that are compliant to the constraints in tracking region

### PR validation:

Results for CMSSW_11_2_0_pre2 , RelValZMM_14

A brief overview :

RvR := Regionally reconstructed tracks passing the selector
GvR := Global pixel tracks passing the selector

####  SUMMARY for TrackingRegion  :  hltIterL3MuonPixelTracksTrackingRegions, TrackCollection  :  hltIterL3MuonPixelTracks
* Total number of events triggered evts    :    15301
* Regionally reconstructed tracks          :    16852
* Total RvR Tracks                         :    16256
* Missed RegReco Tracks in  RvR Tracks     :    596
* Total GvR Tracks                         :    16331
* Matched RegReco Tracks in  GvR Tracks    :    16244
* Missed RegReco Tracks in  GvR Tracks     :    608
* Addtional Tracks in GvR                  :    87

####  SUMMARY for TrackingRegion  :  hltPixelTracksTrackingRegionsL3MuonNoVtx, TrackCollection  :  hltPixelTracksL3MuonNoVtx
* Total number of events triggered evts    :    8242
* Regionally reconstructed tracks          :    216052
* Total RvR Tracks                         :    126371
* Missed RegReco Tracks in  RvR Tracks     :    89681
* Total GvR Tracks                         :    126269
* Matched RegReco Tracks in  GvR Tracks    :    125859
* Missed RegReco Tracks in  GvR Tracks     :    90193
* Addtional Tracks in GvR                  :    410

Plots for the same are attached alongside 
![TrackSelectorByRegion(3)](https://user-images.githubusercontent.com/32837065/90008355-d04d2a80-dcb9-11ea-85c7-d4bb846f2559.png)
![TrackSelectorByRegion(2)](https://user-images.githubusercontent.com/32837065/90008360-d216ee00-dcb9-11ea-93b7-dcf2f3b6a094.png)
![TrackSelectorByRegion(1)](https://user-images.githubusercontent.com/32837065/90008361-d2af8480-dcb9-11ea-88e9-077f5c151873.png)
![TrackSelectorByRegion](https://user-images.githubusercontent.com/32837065/90008364-d3e0b180-dcb9-11ea-8b1e-af754800bb09.png)

ats2008 committed:
Adding with minimal debugging comments
	new file:   ../plugins/TrackSelectorByRaegion.cc
	new files:   trackSelector_validation/*
	modified:   ../../TkTrackingRegions/interface/GlobalTrackingRegion.h
	modified:   ../../TkTrackingRegions/interface/RectangularEtaPhiTrackingRegion.h
	modified:   ../../TkTrackingRegions/interface/TrackingRegion.h
	modified:   ../../TkTrackingRegions/src/GlobalTrackingRegion.cc
	modified:   ../../TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc

formated the code
unnecesry comments/debug stuff removed

Added the configuraration parameter to drop the track collection (ie. just produce mask)

Update RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc

Co-authored-by: Andrea Bocci <andrea.bocci@cern.ch>

 Changes to be committed:
	modified:   plugins/TrackSelectorByRegion.cc
	modified:   ../TkTrackingRegions/interface/GlobalTrackingRegion.h
	modified:   ../TkTrackingRegions/interface/RectangularEtaPhiTrackingRegion.h

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17564

- This PR adds an extra 624KB to repository

- Found files with invalid states:
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/dqmPlots.py:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/validate.cc:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/Harvesting_cfg.py:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/analyze.py:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/IsoMu24_config.py:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17564/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17564/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+  public:
+    explicit TrackSelectorByRegion(const edm::ParameterSet& conf)
+        : tracksToken_(consumes<reco::TrackCollection>(conf.getParameter<edm::InputTag>("tracks"))) {
+      inputTrkRegionToken_ = consumes<edm::OwnVector<TrackingRegion>>(conf.getParameter<edm::InputTag>("inputROI"));
```
I would suggest to rename `"inputROI"` to `"regions"`

mtosi commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+    explicit TrackSelectorByRegion(const edm::ParameterSet& conf)
+        : tracksToken_(consumes<reco::TrackCollection>(conf.getParameter<edm::InputTag>("tracks"))) {
+      inputTrkRegionToken_ = consumes<edm::OwnVector<TrackingRegion>>(conf.getParameter<edm::InputTag>("inputROI"));
+      produce_collection = conf.getParameter<bool>("produceTrackCollection");
```
as briefly discussed today, it might be useful to have an other boolean parameter for handling the mask collection

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+
+      auto tracksHandle = iEvent.getHandle(tracksToken_);
+
+      if (tracksHandle.isValid()) {
```
I would remove the `isValid()` check; otherwise this module sill silently ignore any errors in the configuration.
If the handle is not valid it's OK to get a run time error.

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+      auto tracksHandle = iEvent.getHandle(tracksToken_);
+
+      if (tracksHandle.isValid()) {
+        const auto tracks = *tracksHandle;
```
```suggestion
        const auto& tracks = *tracksHandle;
```
this should be a reference, otherwise the tracks are needlessly copied

mtosi commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+          const auto& regions = *regionsHandle;
+
+          for (const auto& tmp : regions)
+            if (const auto* roi = dynamic_cast<const TrackingRegion*>(&tmp)) {
```
the dynamic_cast should not be needed any longer, because you have introduced the function in the TrackingRegion itself, now
(each region has its own implementation of the function, and it will be automatically called)

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+      if (tracksHandle.isValid()) {
+        const auto tracks = *tracksHandle;
+        mask->assign(tracks.size(), false);
+        if (regionsHandle.isValid()) {
```
I would remove this check - see above

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+        if (regionsHandle.isValid()) {
+          const auto& regions = *regionsHandle;
+
+          for (const auto& tmp : regions)
```
I would rename `tmp` to `region`

mtosi commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+            if (const auto* roi = dynamic_cast<const TrackingRegion*>(&tmp)) {
+              auto amask = roi->checkTracks(tracks);
+              for (size_t it = 0; it < amask.size(); it++) {
+                mask->at(it) = mask->at(it) or amask.at(it);
```
why don't you simply have
`mask->at(it) = amask.at(it)` ?

fwyzard replied:
no, I think the original code is correct - otherwise you simply get the selection from the last region

mtosi replied:
right, thanks !

mtosi commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+        assert(mask->size() == tracks.size());
+
+        for (size_t it = 0; it < mask->size(); it++)
+          if (mask->at(it))
```
I would move this in the above loop, in order to avoid to have 2 loops

fwyzard replied:
Actually I think it's better to have a separate loop: if you do in the loop above, you need to check that a track is not inserted multiple times if it matches multiple regions.

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+
+          for (const auto& tmp : regions)
+            if (const auto* roi = dynamic_cast<const TrackingRegion*>(&tmp)) {
+              auto amask = roi->checkTracks(tracks);
```
can be `auto const& ...` to avoid a copy (even though the compiler should optimise it away here)

mtosi commented on RecoTracker/TkTrackingRegions/interface/GlobalTrackingRegion.h:
```diff
@@ -51,6 +51,7 @@ class GlobalTrackingRegion final : public TrackingRegion {
                                               float gz = 0,
                                               float dr = 0,
                                               float dz = 0) const override;
+  std::vector<bool> checkTracks(reco::TrackCollection const& InputCollection) const override;
```
do you mind to add the comment above this function as you have in the TrackingRegion.h ?
`/// return a mask over a track collection reflecting the compatability to the region`

mtosi commented on RecoTracker/TkTrackingRegions/interface/RectangularEtaPhiTrackingRegion.h:
```diff
@@ -175,6 +175,8 @@ class RectangularEtaPhiTrackingRegion final : public TrackingRegion {
 
   TrackingRegion::Hits hits(const edm::EventSetup& es, const SeedingLayerSetsHits::SeedingLayer& layer) const override;
 
+  std::vector<bool> checkTracks(reco::TrackCollection const& InputCollection) const override;
```
do you mind to add the comment above this function as you have in the TrackingRegion.h ?
`/// return a mask over a track collection reflecting the compatability to the region`

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+          for (const auto& tmp : regions)
+            if (const auto* roi = dynamic_cast<const TrackingRegion*>(&tmp)) {
+              auto amask = roi->checkTracks(tracks);
+              for (size_t it = 0; it < amask.size(); it++) {
```
I would suggest to use `i` for the counter instead of `it`; `it` looks like an iterator :-)

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+
+        assert(mask->size() == tracks.size());
+
+        for (size_t it = 0; it < mask->size(); it++)
+          if (mask->at(it))
+            output_tracks->push_back(tracks[it]);
```
these should be run only `if(produce_collection) { ... }`

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,85 @@
+#include <vector>
+#include <memory>
+
+namespace {
```
Why do toy put everything in an anonymous namespace ?

It's probably correct, but I don't think I have seen it used frequently (or at all) ?

fwyzard commented on RecoTracker/TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc:
```diff
@@ -47,6 +47,34 @@ namespace {
+      continue;
+    }
+
+    if (!etaRange().inside(track.eta())) {
```
if I understand the original code correctly, this (and the next) check should be applied only if `theUseEtaPhi == true` ?

fwyzard replied:
OK, not, that parameter does something different ...

fwyzard commented on RecoTracker/TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc:
```diff
@@ -47,6 +47,34 @@ namespace {
+    if (!etaRange().inside(track.eta())) {
+      continue;
+    }
+    if (std::abs(reco::deltaPhi(track.phi(), phiDirection())) > phiMargin().right()) {
```
see above

fwyzard replied:
also, since `phiMargin()` could potentially be asymmetric, should this make explicit checks for direction - margin and direction + margin ?

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17595

- This PR adds an extra 16KB to repository

- Found files with invalid states:
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/dqmPlots.py:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/validate.cc:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/Harvesting_cfg.py:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/analyze.py:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb
  - RecoTracker/FinalTrackSelectors/test/trackSelector_validation/IsoMu24_config.py:
    - Added: 9cc883601cc867ed2547c2f74f6c3d5ed00f2490
    - Deleted: 33c4b00bcc9cf8d482ce99a72abf2dff6f6f1afb


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17595/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17595/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

slava77 commented:
> Code check has found code style and quality issues which could be resolved by applying following patch(s)

this should be addressed before further tests can run

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17612

- This PR adds an extra 20KB to repository

cmsbuild commented:
A new Pull Request was created by  @ats2008 (Aravind Sugunan) for master.

It involves the following packages:

RecoTracker/FinalTrackSelectors
RecoTracker/TkTrackingRegions


@perrotta, @jpata, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo, @GiacomoSguazzoni, @JanFSchulte, @rovere, @VinInn, @mschrode, @gpetruc, @ebrondol, @mtosi, @dgulhan this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,83 @@
+  using MaskCollection = std::vector<bool>;
+
+  void produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup&) const override {
+    if (not produce_collection)
+      if (not produce_mask)
+        return;
```
```suggestion
    if (not produce_collection and not produce_mask)
      return;

```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,83 @@
+    mask->assign(tracks.size(), false);
+    const auto& regions = *regionsHandle;
+
+    for (const auto& region : regions)
+      if (const auto* roi = dynamic_cast<const TrackingRegion*>(&region)) {
+        auto amask = roi->checkTracks(tracks);
+        for (size_t it = 0; it < amask.size(); it++) {
+          mask->at(it) = mask->at(it) or amask.at(it);
+        }
+      }
```
```suggestion
    for (const auto& region : regions) {
      auto const& region_mask = region.checkTracks(tracks);
      for (size_t i = 0; i < region_mask.size(); ++i) {
        mask->at(i) = mask->at(i) or region_mask[i];
      }
    }
```

fwyzard replied:
the `dynamic_cast` is no longer needed

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,83 @@
+    if (produce_collection)
+      iEvent.put(std::move(output_tracks));
+  }
+  bool produce_collection;
+  bool produce_mask;
+  edm::EDGetTokenT<reco::TrackCollection> tracksToken_;
+  edm::EDGetTokenT<edm::OwnVector<TrackingRegion>> inputTrkRegionToken_;
```
```suggestion

  const bool produce_collection;
  const bool produce_mask;
  const edm::EDGetTokenT<reco::TrackCollection> tracksToken_;
  const edm::EDGetTokenT<edm::OwnVector<TrackingRegion>> inputTrkRegionToken_;
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,83 @@
+        }
+      }
+
+    if (produce_collection)
+      for (size_t it = 0; it < mask->size(); it++)
+        if (mask->at(it))
+          output_tracks->push_back(tracks[it]);
+
+    if (produce_mask)
+      iEvent.put(std::move(mask));
+    if (produce_collection)
+      iEvent.put(std::move(output_tracks));
```
```suggestion
    if (produce_mask) {
      iEvent.put(std::move(mask));
    }

    if (produce_collection) {
      auto output_tracks = std::make_unique<reco::TrackCollection>();  // selected output collection
      size_t size = 0;
      for (size_t i = 0; i < mask->size(); i++) {
        if (mask[i])
          ++size;
      }
      output_tracks->reserve(size);

      for (size_t i = 0; i < mask->size(); i++) {
        if (mask[i])
          output_tracks->push_back(tracks[i]);
      }
      iEvent.put(std::move(output_tracks));
    }
```

fwyzard replied:
This allocates the `output_tracks` collection only if it's actually needed, and reserves the correct size to avoid reallocations.

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,83 @@
+        return;
+    // products
+    auto mask = std::make_unique<MaskCollection>();                  // mask w/ the same size of the input collection
+    auto output_tracks = std::make_unique<reco::TrackCollection>();  // selected output collection
```
```suggestion
```
Moved below.

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,83 @@
+    auto tracksHandle = iEvent.getHandle(tracksToken_);
+
+    const auto& tracks = *tracksHandle;
+    mask->assign(tracks.size(), false);
```
```suggestion
    auto mask = std::make_unique<MaskCollection>(tracks.size(), false);
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,83 @@
+    if (not produce_collection)
+      if (not produce_mask)
+        return;
+    // products
+    auto mask = std::make_unique<MaskCollection>();                  // mask w/ the same size of the input collection
+    auto output_tracks = std::make_unique<reco::TrackCollection>();  // selected output collection
+
```
```suggestion
```

Moved below.

fwyzard commented:
I've suggested some minor (but extensive, I guess) code clean up.

ats2008 committed:
Apply suggestions from code review

committing updates and suggestion from Andrea

Co-authored-by: Andrea Bocci <andrea.bocci@cern.ch>

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

ERROR: Build errors found during clang-tidy run.
```
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:19:12: error: constructor for 'TrackSelectorByRegion' must explicitly initialize the const member 'produce_collection' [clang-diagnostic-error]
  explicit TrackSelectorByRegion(const edm::ParameterSet& conf)
           ^
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:81:14: note: declared here
--
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:19:12: error: constructor for 'TrackSelectorByRegion' must explicitly initialize the const member 'produce_mask' [clang-diagnostic-error]
  explicit TrackSelectorByRegion(const edm::ParameterSet& conf)
           ^
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:82:14: note: declared here
--
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:21:26: error: no viable overloaded '=' [clang-diagnostic-error]
    inputTrkRegionToken_ = consumes<edm::OwnVector<TrackingRegion>>(conf.getParameter<edm::InputTag>("regions"));
                         ^
/cvmfs/cms-ib.cern.ch/nweek-02640/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-06-1100/src/FWCore/Utilities/interface/EDGetToken.h:61:9: note: candidate function (the implicit copy assignment operator) not viable: 'this' argument has type 'const edm::EDGetTokenT<edm::OwnVector<TrackingRegion> >', but method is not marked const
--
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:22:24: error: cannot assign to non-static data member 'produce_collection' with const-qualified type 'const bool' [clang-diagnostic-error]
    produce_collection = conf.getParameter<bool>("produceTrackCollection");
                       ^
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:81:14: note: non-static data member 'produce_collection' declared const here
--
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:23:18: error: cannot assign to non-static data member 'produce_mask' with const-qualified type 'const bool' [clang-diagnostic-error]
    produce_mask = conf.getParameter<bool>("produceMask");
                 ^
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:82:14: note: non-static data member 'produce_mask' declared const here
--
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:68:17: error: type 'std::unique_ptr<std::vector<bool, std::allocator<bool> >, std::default_delete<std::vector<bool, std::allocator<bool> > > >' does not provide a subscript operator [clang-diagnostic-error]
        if (mask[i])
                ^
RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:74:17: error: type 'std::unique_ptr<std::vector<bool, std::allocator<bool> >, std::default_delete<std::vector<bool, std::allocator<bool> > > >' does not provide a subscript operator [clang-diagnostic-error]
        if (mask[i])
                ^
Suppressed 1497 warnings (1496 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

ats2008 committed:
Added assymetric phi margin
	modified:   RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc
	modified:   RecoTracker/TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17632

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #31055 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

fwyzard commented:
I have a couple of technical questions:
  - looks like [`boost::dynamic_bitset`](https://www.boost.org/doc/libs/1_72_0/libs/dynamic_bitset/dynamic_bitset.html) would be more appropriate than `std::vector<bool>` for the mask; would there be any downsides from using it?
  - what is the efficient way of putting a collection in the event ? would it be better to allocate the collection on the stack instead of with `make_unique`, and then move it into the event with `iEvent.emplace` instead of `iEvent.put` ?

makortel commented:
> * what is the efficient way of putting a collection in the event ? would it be better to allocate the collection on the stack instead of with `make_unique`, and then move it into the event with `iEvent.emplace` instead of `iEvent.put` ?

`Event::emplace()` is a tiny bit more efficient because it avoids the `std::unique_ptr` (and thus one heap allocation for the collection object).

(`Event::put()` also does a move-assignment from the data product from the argument `std::unique_ptr` to an object inside the `edm::Wrapper`)

VinInn commented:
what would be the advantage of the boost container?
in any case the CMSSW core team developed ContainerMask https://cmssdt.cern.ch/dxr/CMSSW/source/DataFormats/Common/interface/ContainerMask.h

fwyzard commented:
> what would be the advantage of the boost container?

At least two things that we need to do are already implemented:
  - container-wide logical and
  - count of the number bits set 

> in any case the CMSSW core team developed ContainerMask https://cmssdt.cern.ch/dxr/CMSSW/source/DataFormats/Common/interface/ContainerMask.h

Mhm, that's a `RefProd` plus the `vector<bool>` ?
I guess that could make sense as the object put in the Event, as it would keep track of what the mask refers to (the original track collection).

So, should we use
  - `boost::dynamic_bitset` as the return type of `TrackingRegion::checkTracks()` and for the intermediate state inside `TrackSelectorByRegion`
  - `edm::ContainerMask<reco::TrackCollection>` for the product going into the event

?

VinInn commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,87 @@
+      }
+      output_tracks->reserve(size);
+      for (size_t i = 0; i < mask->size(); i++) {
+        if (mask->at(i))
```
avoid at

VinInn replied:
More in general: can we stop copying objects by value from one collection to another and start consuming edm::View?

fwyzard replied:
Mhm, no, because the whole purpose of this selector is to be a drop-in replacement for the existing regional tracking.
Moving all clients to consume `edm::View` would need somebody to
  - do the work of migrating and validating each client
  - check the the HLT time does not actually increase using `edm::View`

It *can* be done, but 1. it's way beyond the scope of this module, 2. will probably take much longer, 3. can anyway be done in parallel.

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,87 @@
+    for (auto const& region : regions) {
+      auto const& region_mask = region.checkTracks(tracks);
+      for (size_t i = 0; i < region_mask.size(); ++i) {
+        mask->at(i) = mask->at(i) or region_mask[i];
```
> never use at; it checks boundaries and can throw

I agree when the index is clearly within the boundary, like for `region_mask` since the loop is on its size.
In general, I don't mind the extra check, given the number of times I've seen people access data outside the container :-(

Here I'd be fine changing `mask->at[i]` to `(*mask)[i]` only if we add an `assert(mask->size() == regiona_mask.size())` before the loop...

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,87 @@
+    const auto& regions = *regionsHandle;
+
+    for (auto const& region : regions) {
+      auto const& region_mask = region.checkTracks(tracks);
```
Makes sense... I'd probably keep both versions (the inefficient one can be implemented calling the efficient one).

VinInn commented:
> It can be done, but 1. it's way beyond the scope of this module, 2. will probably take much longer, 3. can anyway be done in parallel.

It was a general comment. I agree it is a long term project.

btw the copy here is shallow not deep, Extra and Hits are not copied...

fwyzard commented:
> btw the copy here is shallow not deep, Extra and Hits are not copied...

That's... good, right ?

VinInn commented:
> On 7 Aug, 2020, at 8:00 PM, Andrea Bocci <notifications@github.com> wrote:
> 
> 
> btw the copy here is shallow not deep, Extra and Hits are not copied...
> 
> That's... good, right ?
> 
as long as no one pretends to make early delete based on the track collection....

v.

ats2008 committed:
remover ->at() s
    modified:   plugins/TrackSelectorByRegion.cc

ats2008 committed:
Added the suggested new signature for checkTracks
Updated the TrackSelectorByRegion
    modified:   plugins/TrackSelectorByRegion.cc
	modified:   ../TkTrackingRegions/interface/GlobalTrackingRegion.h
	modified:   ../TkTrackingRegions/interface/RectangularEtaPhiTrackingRegion.h
	modified:   ../TkTrackingRegions/interface/TrackingRegion.h
	modified:   ../TkTrackingRegions/src/GlobalTrackingRegion.cc
	modified:   ../TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17672

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #31055 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,86 @@
+    }
+  }
+
+  const bool produce_collection;
+  const bool produce_mask;
```
these should be renamed to follow the coding rules:
```suggestion
  const bool produceCollection_;
  const bool produceMask_;
```

cmsbuild commented:
The tests are being triggered in jenkins.

fwyzard commented on RecoTracker/TkTrackingRegions/interface/TrackingRegion.h:
```diff
@@ -96,6 +99,14 @@ class TrackingRegion {
   /// get hits from layer compatible with region constraints
   virtual Hits hits(const edm::EventSetup& es, const SeedingLayerSetsHits::SeedingLayer& layer) const = 0;
 
+  /// updates an existing  mask over a track collection reflecting the compatability to the region
```
```suggestion
  /// Set the elements of the mask corresponding to the tracks that are compatable with the region.
  /// Does not reset the elements corresponding to the tracks that are not compatible.
```
also in the other files

fwyzard commented on RecoTracker/TkTrackingRegions/interface/TrackingRegion.h:
```diff
@@ -96,6 +99,14 @@ class TrackingRegion {
+  /// updates an existing  mask over a track collection reflecting the compatability to the region
+  virtual void checkTracks(reco::TrackCollection const& InputCollection, std::vector<bool>& mask) const = 0;
+
+  std::vector<bool> checkTracks(reco::TrackCollection const& InputCollection) const {
```
Add a comment ?

fwyzard commented on RecoTracker/TkTrackingRegions/src/GlobalTrackingRegion.cc:
```diff
@@ -90,28 +88,52 @@ std::unique_ptr<HitRZCompatibility> GlobalTrackingRegion::checkRZ(const DetLayer
+
+void GlobalTrackingRegion::checkTracks(reco::TrackCollection const& InputCollection, std::vector<bool>& mask) const {
+  math::XYZPoint regOrigin(origin().x(), origin().y(), origin().z());
+  if (mask.size() < InputCollection.size())
+    mask.resize(InputCollection.size(), false);
```
I would replace this with
```suggestion
  assert (mask.size() == InputCollection.size());
```

fwyzard commented on RecoTracker/TkTrackingRegions/interface/TrackingRegion.h:
```diff
@@ -96,6 +99,14 @@ class TrackingRegion {
   virtual Hits hits(const edm::EventSetup& es, const SeedingLayerSetsHits::SeedingLayer& layer) const = 0;
 
+  /// updates an existing  mask over a track collection reflecting the compatability to the region
+  virtual void checkTracks(reco::TrackCollection const& InputCollection, std::vector<bool>& mask) const = 0;
```
`InputCollection` should be lower case; if it doesn't cause conflicts, I would suggest to use `tracks` everywhere.

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,86 @@
+
+    if (produce_collection) {
+      auto output_tracks = std::make_unique<reco::TrackCollection>();  // selected output collection
+      size_t size = 0;
+
```
```suggestion
      // count the number of selected tracks
      size_t size = 0;
```

cmsbuild commented:
-1

Tested at: d9114947bbcf8b2b3e492e55e9f6bd3ebf802e38

CMSSW: CMSSW_11_2_X_2020-08-10-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0ab8c1/8689/summary.html

I found follow errors while testing this PR

**Failed tests:  Build**

* **Build**:

I found compilation error when building: 

<pre>                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-08-10-1100/src/RecoTracker/TkTrackingRegions/interface/TrackingRegionBase.h:1,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-08-10-1100/src/RecoTracker/SpecialSeedGenerators/interface/CosmicTrackingRegion.h:11,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-08-10-1100/src/RecoTracker/SpecialSeedGenerators/src/CosmicTrackingRegion.cc:1:
/cvmfs/cms-ib.cern.ch/nweek-02641/slc7_amd64_gcc820/external/gcc/8.2.0-bcolbf/include/c++/8.4.0/bits/unique_ptr.h: In instantiation of 'typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...) [with _Tp = CosmicTrackingRegion; _Args = {const CosmicTrackingRegion&}; typename std::_MakeUniq<_Tp>::__single_object = std::unique_ptr<CosmicTrackingRegion, std::default_delete<CosmicTrackingRegion> >]':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-08-10-1100/src/RecoTracker/SpecialSeedGenerators/interface/CosmicTrackingRegion.h:97:111:   required from here
/cvmfs/cms-ib.cern.ch/nweek-02641/slc7_amd64_gcc820/external/gcc/8.2.0-bcolbf/include/c++/8.4.0/bits/unique_ptr.h:835:30: error: invalid new-expression of abstract class type 'CosmicTrackingRegion'
     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-08-10-1100/src/RecoTracker/SpecialSeedGenerators/src/CosmicTrackingRegion.cc:1:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-08-10-1100/src/RecoTracker/SpecialSeedGenerators/interface/CosmicTrackingRegion.h:24:7: note:   because the following virtual functions are pure within 'CosmicTrackingRegion':
 class CosmicTrackingRegion : public TrackingRegionBase {
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

ats2008 committed:
Added Cosmic Tracking Region
Added suggestions by @andrea.bocci
Added comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17679

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #31055 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,84 @@
+        if ((*mask)[i])
+          output_tracks->push_back(tracks[i]);
+      }
+      iEvent.emplace(outputTracksToken_, *output_tracks);
```
```suggestion
      iEvent.emplace(outputTracksToken_, std::move(output_tracks));
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    if (produceCollection_) {
+      auto output_tracks = std::make_unique<reco::TrackCollection>();  // selected output collection
```
```suggestion
      reco::TrackCollection output_tracks;  // output collection with a (shallow) copy of the selected tracks
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,84 @@
+      for (size_t i = 0; i < mask->size(); i++) {
+        size += (*mask)[i];
+      }
+      output_tracks->reserve(size);
```
```suggestion
      output_tracks.reserve(size);
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,84 @@
+      output_tracks->reserve(size);
+      for (size_t i = 0; i < mask->size(); i++) {
+        if ((*mask)[i])
+          output_tracks->push_back(tracks[i]);
```
```suggestion
          output_tracks.push_back(tracks[i]);
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,84 @@
+    if (not produceCollection_ and not produceMask_)
+      return;
+
+    auto regionsHandle = iEvent.getHandle(inputTrkRegionToken_);
+    auto tracksHandle = iEvent.getHandle(tracksToken_);
+
+    const auto& tracks = *tracksHandle;
+    auto mask = std::make_unique<MaskCollection>(tracks.size(), false);  // output mask
+    const auto& regions = *regionsHandle;
```
```suggestion
    const auto& regions = *regionsHandle;
    const auto& tracks = *tracksHandle;
    MaskCollection mask(tracks.size(), false);  // output mask
``````

fwyzard replied:
Even simpler:
```suggestion
    auto const& regions = iEvent.get(inputTrkRegionToken_);
    auto const& tracks = iEvent.get(tracksToken_);
    MaskCollection mask(tracks.size(), false);  // output mask
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,84 @@
+    const auto& regions = *regionsHandle;
+
+    for (auto const& region : regions) {
+      region.checkTracks(tracks, *mask);
```
```suggestion
      region.checkTracks(tracks, mask);
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,84 @@
+      auto output_tracks = std::make_unique<reco::TrackCollection>();  // selected output collection
+      size_t size = 0;
+      // count the number of selected tracks
+      for (size_t i = 0; i < mask->size(); i++) {
+        size += (*mask)[i];
```
```suggestion
      for (size_t i = 0; i < mask.size(); i++) {
        size += mask[i];
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,84 @@
+        size += (*mask)[i];
+      }
+      output_tracks->reserve(size);
+      for (size_t i = 0; i < mask->size(); i++) {
+        if ((*mask)[i])
```
```suggestion
      for (size_t i = 0; i < mask.size(); i++) {
        if (mask[i])
```

fwyzard commented on RecoTracker/FinalTrackSelectors/plugins/TrackSelectorByRegion.cc:
```diff
@@ -0,0 +1,84 @@
+      iEvent.emplace(outputTracksToken_, *output_tracks);
+    }
+    if (produceMask_) {
+      iEvent.emplace(outputMaskToken_, *mask);
```
```suggestion
      iEvent.emplace(outputMaskToken_, std::move(mask));
```

fwyzard reviewed (changes requested):
The idea of using `Event::emplace(...)` is that you don't need to allocate a `unique_ptr`; you can simply allocate the collection on the stack and then "move" it into the event.

fwyzard commented on RecoTracker/TkTrackingRegions/interface/TrackingRegion.h:
```diff
@@ -96,6 +99,16 @@ class TrackingRegion {
+  /// Does not reset the elements corresponding to the tracks that are not compatible.
+  virtual void checkTracks(reco::TrackCollection const& tracks, std::vector<bool>& mask) const = 0;
+
+  /// retutns a boolean mask over the TrackCollection reflecting the comapatbility of each track with the region constraints
```
```suggestion
  /// return a boolean mask over the TrackCollection reflecting the compatibility of each track with the region constraints
```

ats2008 committed:
Apply suggestions from code review

Some more modifications and optimizations

Co-authored-by: Andrea Bocci <andrea.bocci@cern.ch>

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17686

- This PR adds an extra 40KB to repository

cmsbuild commented:
Pull request #31055 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

fwyzard commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f780524a8b0309f731264768a787466944ddd683
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0ab8c1/8695/summary.html
CMSSW: CMSSW_11_2_X_2020-08-10-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

fwyzard commented:
@slava77 @perrotta @jpata I think we are finally done with the updates :-)
Could you review the PR ?

@mtosi @vmariani do we need a backport to 11.1.x, *e.g.* for the Phase 2 HLT ?

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0ab8c1/8695/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612401
 - DQMHistoTests: Total failures: 11
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2612341
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

mtosi commented:
yes, a backport to 111X for the phase2 HLT TDR studies would be indeed useful

jpata commented:
Thanks for the review so far!

@ats2008 do you have a test for the new EDProducer `TrackSelectorByRegion`? The current workflows are not modified by the changes in the existing modules [1] and the code looks fine to me, but it would be good to test the new code as well.

[1] https://github.com/cms-sw/cmssw/pull/31055#issuecomment-671869969

fwyzard commented:
@jpata do you mean a link to results of using the new module, or do you mean adding a test workflow in CMSSW ?

For the latter, is there any test workflow that makes use of regional tracking one could start from ?

jpata commented:
> @jpata do you mean a link to results of using the new module, or do you mean adding a test workflow in CMSSW ?

@fwyzard I was somewhat vague here indeed. Showing the testing procedure and the results would already be useful!
 
> For the latter, is there any test workflow that makes use of regional tracking one could start from ?

I'm looking through the code to see if I find one, but it's possible that tracking experts can answer that faster.

fwyzard commented:
> Showing the testing procedure and the results would already be useful!

Aravind presented the results (obtained from a previous version of the code) at the last TSG meeting: [TrackSelectorByRegion.pdf](https://indico.cern.ch/event/939462/contributions/3969981/attachments/2084986/3502599/TrackSelectorByRegion.pdf)

ats2008 commented:
@jpata @fwyzard I have updated the PR description with the plots/summary for the CMSSW_11_2_0_pre2 , RelValZMM_14 with the most recent code

jpata commented on RecoTracker/SpecialSeedGenerators/src/CosmicTrackingRegion.cc:
```diff
@@ -24,6 +24,30 @@ namespace {
 using namespace std;
 
+void CosmicTrackingRegion::checkTracks(reco::TrackCollection const& tracks, std::vector<bool>& mask) const {
+  math::XYZPoint regOrigin(origin().x(), origin().y(), origin().z());
```
make it const

jpata commented on RecoTracker/TkTrackingRegions/src/GlobalTrackingRegion.cc:
```diff
@@ -90,28 +88,51 @@ std::unique_ptr<HitRZCompatibility> GlobalTrackingRegion::checkRZ(const DetLayer
 }
+
+void GlobalTrackingRegion::checkTracks(reco::TrackCollection const& tracks, std::vector<bool>& mask) const {
+  math::XYZPoint regOrigin(origin().x(), origin().y(), origin().z());
```
make it const

fwyzard replied:
does it make any difference ?

jpata replied:
I haven't checked if the generated code is different, and I don't expect it would be. However, my understanding is that we can and should be liberal with `const` in signaling intent to the compiler. Let me know if I have misunderstood. Thanks!

fwyzard replied:
Adding `const` clearly doesn't harm - but as you say, I don't expect to be any difference in the generated code, since in simple cases like this the compiler should figure out that the variable(s) are not being modified.
I'd say it's more about signalling intention to the developers, which of course doesn't harm...

jpata commented on RecoTracker/TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc:
```diff
@@ -47,6 +48,37 @@ namespace {
 using namespace std;
 
+void RectangularEtaPhiTrackingRegion::checkTracks(reco::TrackCollection const& tracks, std::vector<bool>& mask) const {
+  math::XYZPoint regOrigin(origin().x(), origin().y(), origin().z());
```
make all these const

ats2008 committed:
Added const ness to variables. from@jpata

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31055/17712

- This PR adds an extra 40KB to repository

cmsbuild commented:
Pull request #31055 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

Thanks for the update! I'll just run a quick build and then it should be fine from my side.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 2888bd94dacc34a696e1826ccd1ccfa958ed0582
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0ab8c1/8726/summary.html
CMSSW: CMSSW_11_2_X_2020-08-12-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0ab8c1/8726/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612401
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2612347
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jpata commented:
+1

- tests pass, changes to existing modules do not modify reco output as far as we can tell
- new `TrackSelectorByRegion` EDProducer has been implemented and presented at the TSG
- requested changes have been implemented
- per https://github.com/cms-sw/cmssw/pull/31055#issuecomment-671899619 we expect a backport to 11_1_X for phase 2 HLT TDR studies

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
